### PR TITLE
Add os-release file parsing, partially fix #25897

### LINKED ISSUE #25897: Add Atomic OS detection in Ansible Facts
##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
facts.py

##### ANSIBLE VERSION
N/A

##### SUMMARY
Atomic OS (http://www.projectatomic.io/) isn't detected in Ansible Fact.

Maybe set:

```
"ansible_os_family": "Atomic RedHat"
```

### PR DESCRIPTION
**Pull request by mscherer**
##### SUMMARY

Add os-release file parsing, partially fix #25897
/etc/os_release is documented on https://www.freedesktop.org/software/systemd/man/os-release.html

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
setup

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.5.0 (add_fact_os_release 1b2081dfbb) last updated 2017/12/12 15:56:51 (GMT +200)
  config file = /home/misc/.ansible.cfg
  configured module search path = [u'/home/misc/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/misc/checkout/git/ansible/lib/ansible
  executable location = /home/misc/checkout/git/ansible/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

### CODE REVIEW

**Comment by mscherer** [comment]
I didn't took a look at testing that yet, but plan to add it to the unit tests.

File: lib/ansible/module_utils/facts/system/os_release.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            if line.startswith('#'):
+                continue
+            for t in shlex.split(line):
+                (k, v) = t.split('=', 1)
+                result[k] = v
+        return result
+
+    def collect(self, module=None, collected_facts=None):
+        facts_dict = {}
+        path = None
```
Comment (pilou-) [inline lib/ansible/module_utils/facts/system/os_release.py:44]: A comment about `The file /etc/os-release takes precedence over /usr/lib/os-release.` could be added there.

File: test/units/module_utils/facts/test_collectors.py [hunk 2]
```diff
@@ -218,6 +219,23 @@ class TestNetworkCollector(BaseFactsTest):
...
 
 
+class TestOSReleaseFacts(BaseFactsTest):
+    __test__ = True
+    gather_subset = ['!all', 'os_release']
+    valid_subsets = ['os_release']
+    fact_namespace = 'ansible_os_release'
+    collector_class = OSReleaseFactCollector
+
+    @patch('ansible.module_utils.facts.system.os_release.get_file_content', return_value='# test\nID=clown_linux')
```
Comment (pilou-) [inline test/units/module_utils/facts/test_collectors.py]: Another variable with value using `special characters outside of A–Z, a–z, 0–9` and enclosed with quotes might be added there.

File: lib/ansible/module_utils/facts/system/os_release.py (hunk 1 above)
Comment (mscherer) [reply lib/ansible/module_utils/facts/system/os_release.py:44]: Done

File: test/units/module_utils/facts/test_collectors.py (hunk 2 above)
Comment (mscherer) [reply test/units/module_utils/facts/test_collectors.py]: I added quotes, not sure about special char.

Comment (pilou-) [reply test/units/module_utils/facts/test_collectors.py]: A space is good enough :)

**Comment by pilou-** [comment]
Although it doesn't count: shipit .

File: lib/ansible/module_utils/facts/system/os_release.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from ansible.module_utils.facts.collector import BaseFactCollector
+
+
+class OSReleaseFactCollector(BaseFactCollector):
+    name = 'os-release'
+    _fact_ids = set(['os_release'])
+
+    def parse_file(self, path):
+        result = {}
+        for line in get_file_content(path).splitlines():
```
Comment (alikins) [inline lib/ansible/module_utils/facts/system/os_release.py]: I would split this method into:
- reading in the file content to a string
- parsing the string

Then the test cases can test the parsing without having to patch get_file_content or any file io.

May be easier to test if the parsing function was module scope method as well, then the parsing tests dont
need to construct a collector or a faux AnsibleModule. The OsReleaseFactCollector would still be useful for
adding tests for existence of /etc/os-release and general class tests.

File: lib/ansible/module_utils/facts/system/os_release.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            if line.startswith('#'):
+                continue
+            for t in shlex.split(line):
```
Comment (alikins) [inline lib/ansible/module_utils/facts/system/os_release.py]: One odd thing that has come up with the os-release parsing code in module_utils/facts/system/distribution.py is the quoting of some of the right hand side of lines. 

I'll track down the bug momentarily, but at least one distro did odd things about what was/wasn't quoted  (either VERSION wasnt quoted or ID was quoted?). The lsb/os-release spec is kind of vague on that, and different '/bin/lsb_release' tools seemed to do different things.

This isn't something that has to be addressed since I don't think there was a clear correct answer, but just mentioning it for context.

File: test/units/module_utils/facts/test_collectors.py [hunk 5]
```diff
@@ -218,6 +219,24 @@ class TestNetworkCollector(BaseFactsTest):
...
 
 
+class TestOSReleaseFacts(BaseFactsTest):
+    __test__ = True
+    gather_subset = ['!all', 'os_release']
+    valid_subsets = ['os_release']
+    fact_namespace = 'ansible_os_release'
+    collector_class = OSReleaseFactCollector
+
+    @patch('ansible.module_utils.facts.system.os_release.get_file_content', return_value='# test\nID=clown_linux\nVERSION="Clown Linux"')
```
Comment (alikins) [inline test/units/module_utils/facts/test_collectors.py]: I'd like to see more tests of the parsing, ideally with some examples of real os-release files. test/units/module_utils/test_distribtution.py[1] has some.

[1] test_distribution.py file is in the wrong place but nonetheless

**Review by alikins — DISMISSED** [review]
I'd like to see the parse function be pulled out of the class. And for it to take a string for input and tests updated to use it.

**Comment by alikins** [comment]
@mscherer https://github.com/ansible/ansible/issues/25725 and https://github.com/ansible/ansible/pull/31143 are the issue/pr related to the quoting in /etc/os-release I mentioned

**Comment by alikins** [comment]
@mscherer Though I guess #25725 and #31143 were actually about  /etc/lsb-release not /etc/os-release so may not be relevant.

**Review by samdoran — CHANGES_REQUESTED** [review]
Could change the license of this file to BSD? Files under `lib/ansible/module_utils` be licensed this way. You can look at `basic.py` for an example.

File: lib/ansible/module_utils/facts/system/os_release.py (hunk 4 above)
Comment (mscherer) [reply lib/ansible/module_utils/facts/system/os_release.py]: So, while lsb is vague on that, I think, but the os-release specs is clear:
"Variable assignment values must be enclosed in double or single quotes if they include spaces, semicolons or other special characters outside of A–Z, a–z, 0–9. Shell special characters ("$", quotes, backslash, backtick) must be escaped with backslashes, following shell style."

Now of course, being on the right side of the specs do not help users :/

**Comment by mscherer** [comment]
All others files in facts/system are under GPL:
```
module_utils/facts/system $ ls *py |wc -l
19
module_utils/facts/system $ grep gnu.org *py|wc -l
18
module_utils/facts/system $ cat __init__.py
module_utils/facts/system $ 
```

And I heavily borrowed code and structure from there.

**Comment by mscherer** [comment]
@alikins there is already a set of os-release file in test_distribution, so I either have to import that file in the test_os_release.py I am writing, or place the tests in test_distribution. Any preference ?

The 2nd do not look correct (different modules), the first mean that test have a dependencies, and I am not sure that's correct.

Or would it be better to just duplicate part of the data (I am leaning to that, since I also have to get the result). Would 1 file be enough, provided we test enough things ? (eg, a regular file, then corner case)

**Comment by alikins** [comment]
> @alikins there is already a set of os-release file in test_distribution, so I either have to import that file in the test_os_release.py I am writing, or place the tests in test_distribution. Any preference ?

> The 2nd do not look correct (different modules), the first mean that test have a dependencies, and I am not sure that's correct.

> Or would it be better to just duplicate part of the data (I am leaning to that, since I also have to get the result). Would 1 file be enough, provided we test enough things ? (eg, a regular file, then corner case)

@mscherer I guess my preference would be to go ahead and mv test_distribution into test/units/module_utils/facts/system/test_distribution.py. Quick test that seems like just a 'git mv'

Then either:

1. import test_distribution.TESTSETS into test_os_release 

``` python
from . import test_distribution
```
       (or I guess, more ansible style would be)
``` python
from .test_distribution import TESTSETS
```
       but some variety of that...

2. extract TESTSETS to a separate module or data file and import the testdata module into both

3. A couple of unit tests setup some 'fixtures' data dirs we could try to emulate, but I don't think there is a consistent approach to that yet. Maybe something for the future...

Any of 1,2,3 would be okay with me. I wouldn't worry about trying #3 for this pr though.

**Comment by alikins** [comment]
@mscherer I should mention the main advantage to the test_distributions.TESTSETS at the moment is that hacking/tests/gen_distribution_version_testcase.py  can collect the needed sample info and dump it in the format used by TESTSETS.

That is a pretty simple script though, so nothing that can't be changed if you have a better idea on how to organize the test data.

**Comment by samdoran** [comment]
@mscherer Roger. We're trying to move these to BSD, but if we're stuck with GPL for this then so be it.

**Comment by mattclay** [comment]
CI failure in unit tests due to undefined variable caught by pylint.

**Comment by mscherer** [comment]
@alikins The problem is that TESTSETS do not contain the result of /etc/os_release parsing and if I have to generate it using the same code, I am not sure this is bringing much on the table vs hardcoding the file and the results in a separate data structure.

**Review by samdoran — APPROVED** [review]

**Comment by bcoca** [comment]
Sorry we kept you hanging this long.

We are going to close this as os-release parsing was included in the vendored/updated distribution detection and for the few extra fields we don't return a simple custom inventory plugin can now be added to the default `gather_facts` action to get the same result.

However, we're always up for discussion.

Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed on the page below:

* <https://docs.ansible.com/ansible/latest/community/communication.html>

In the future, sometimes starting a discussion on the Ansible Forum before implementing
a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Comment by bcoca** [comment]
it is just a facts_module, that you can reference via configuration to have the `gather_facts` option use automatically.

https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#creating-an-info-or-a-facts-module

TLDR: retun the info under `ansible_facts` dictionary.
